core/Test/Tasty/Core.hs

```diff
@@ -408,23 +405,57 @@ foldTestTree
      -- ^ the tree to fold
   -> b
 foldTestTree (TreeFold fTest fGroup fResource fAfter) opts0 tree0 =
-  go mempty opts0 tree0
+  go (filterByPattern (evaluateOptions opts0 tree0))
+  where
+    go :: AnnTestTree OptionSet -> b
+    go = \case
+      AnnEmptyTestTree               -> mempty
+      AnnSingleTest opts name test   -> fTest opts name test
+      AnnTestGroup opts name trees   -> fGroup opts name (foldMap go trees)
+      AnnWithResource opts res0 tree -> fResource opts res0 $ \res -> go (tree res)
+      AnnAfter opts deptype dep tree -> fAfter opts deptype dep (go tree)
+
+-- | 'TestTree' with arbitrary annotations, e. g., evaluated 'OptionSet'.
+data AnnTestTree ann
+  = AnnEmptyTestTree
+  | forall t . IsTest t => AnnSingleTest ann TestName t
+  | AnnTestGroup ann TestName [AnnTestTree ann]
+  | forall a . AnnWithResource ann (ResourceSpec a) (IO a -> AnnTestTree ann)
+  | AnnAfter ann DependencyType Expr (AnnTestTree ann)
```

Comment:
Please give a haddock comment for each of the constructors.

If the original `TestTree` has such comments, a minimal doc here would be to link to these original constructors so I can get the explanation by a single click.  But more doc is always welcome!

Reply:
Done, but this data type is not exported any way.